subxt/src/extrinsics/extrinsics_type.rs

```diff
@@ -0,0 +1,344 @@
+        }
+
+        let call_start_idx = extrinsic_bytes.len() - cursor.len();
+
+        // Ensure the provided bytes are sound.
+        scale_decode::visitor::decode_with_visitor(
+            cursor,
+            ids.call,
+            &metadata.runtime_metadata().types,
+            scale_decode::visitor::IgnoreVisitor,
+        )
+        .map_err(scale_decode::Error::from)?;
+
+        // Decode the pallet index, then the call variant.
+        let cursor = &mut &extrinsic_bytes[call_start_idx..];
+        let pallet_index: u8 = Decode::decode(cursor).map_err(Error::Codec)?;
+        let variant_index: u8 = Decode::decode(cursor).map_err(Error::Codec)?;
+
+        Ok(ExtrinsicDetails {
+            index,
+            is_signed,
+            address_start_idx,
+            address_end_idx,
+            call_start_idx,
+            pallet_index,
+            variant_index,
+            bytes: extrinsic_bytes,
+        })
+    }
+
+    /// The index of the extrinsic in the given block.
+    pub fn index(&self) -> usize {
+        self.index
+    }
+
+    /// The index of the pallet that the extrinsic originated from.
+    pub fn pallet_index(&self) -> u8 {
+        self.pallet_index
+    }
+
+    /// The index of the event variant that the event originated from.
+    pub fn variant_index(&self) -> u8 {
+        self.variant_index
+    }
+
+    /// Return _all_ of the bytes representing this extrinsic, which include, in order:
+    /// - First byte: abbbbbbb (a = 0 for unsigned, 1 for signed, b = version)
+    /// - SignatureType (if the payload is signed)
+    ///   - Address
+    ///   - Signature
+    ///   - Extra fields
+    /// - Extrinsic call bytes
+    pub fn bytes(&self) -> &[u8] {
+        &self.bytes
+    }
+
+    /// Return only the bytes representing this extrinsic call.
+    ///
+    /// # Note
+    ///
+    /// Please use `[Self::bytes]` if you want to get all extrinsic bytes.
+    pub fn call_bytes(&self) -> &[u8] {
+        &self.bytes[self.call_start_idx..]
+    }
+
+    /// Return only the bytes of the address that signed this extrinsic.
+    ///
+    /// # Note
+    ///
+    /// Returns `None` if the extrinsic is not signed.
+    pub fn address_bytes(&self) -> Option<&[u8]> {
+        self.is_signed
+            .then(|| &self.bytes[self.address_start_idx..self.address_end_idx])
+    }
+
+    /// Attempt to statically decode the address bytes into the provided type.
+    ///
+    /// # Note
+    ///
+    /// Returns `None` if the extrinsic is not signed.
+    pub fn as_address<T: Decode>(&self) -> Option<Result<T, Error>> {
+        self.address_bytes()
+            .map(|bytes| T::decode(&mut &bytes[..]).map_err(Error::Codec))
+    }
+
+    /// Attempt to statically decode the extrinsic call bytes into the provided type.
+    pub fn as_call<T: Decode>(&self) -> Result<T, Error> {
+        let bytes = &mut &self.call_bytes()[..];
+        T::decode(bytes).map_err(Error::Codec)
+    }
```

Comment:
Can we use `decode_as_fields` instead for this? Since we know the type ID we want to decode as and have metadata?

I'm keen to avoid statically decoding stuff with eg `Decode` wherever possible.

For events, we have a `StaticEvent` type which allows us to point at specific event fields, and know how to decode them then from the metadata. I think this is on your todo list already (but I wouldn't want to merge the `Decode` version basically :))

Also it's not super consistent at the mo because this `as_call` looks like it decodes the entire enum, whereas the event's `as_event` decodes just the event fields (and `as_root_event` decodes the entire enum), so yeah just have a check and make sure that these method names are consistent with the event names when done :)